### PR TITLE
M2P-49 Fix bug - Unable to complete order with virtual item after applying a discount

### PR DESCRIPTION
**Pull request by dumega**
# Description
When trying to apply coupon to virtual qoute on Bolt modal, it would fail, cause the billing address of $immutableQuote is always empty, and the cart data returned from Cart Helper function is empty as a result, this PR is to fix this bug.

Fixes: https://boltpay.atlassian.net/browse/M2P-49

#changelog Fix bug - Unable to complete order with virtual item after applying a discount

# Type of change

- [x] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [x] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?

# Checklist:

- [x] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: Model/Api/DiscountCodeValidation.php [hunk 1]
```diff
@@ -1024,4 +1014,32 @@ protected function getParentQuoteDiscountResult($couponCode, $coupon, $parentQuo
...
+     * @param object $address
+     *
+     * @return array
+     */
+    private function prepareAddressData($address)
+    {
+        $address = $this->cartHelper->handleSpecialAddressCases($address);
+        $region = $this->regionModel->loadByName(@$address->region, @$address->country_code);
+        $addressData = [
+                    'firstname'    => @$address->first_name,
```
Comment (vitaliyreznikov) [inline Model/Api/DiscountCodeValidation.php:1028]: @matt-thomason Do I understand correctly we want to stop use "@" ? Do we have any doc about this?

File: Model/Api/DiscountCodeValidation.php [hunk 2]
```diff
@@ -1024,4 +1014,32 @@ protected function getParentQuoteDiscountResult($couponCode, $coupon, $parentQuo
             'discount_type'   => $this->convertToBoltDiscountType($rule->getSimpleAction()),
         ];
     }
+    
+    /**
+     * @param object $address
+     *
+     * @return array
+     */
+    private function prepareAddressData($address)
```
Comment (vitaliyreznikov) [inline Model/Api/DiscountCodeValidation.php:1023]: @dumega Could you add a comment to this method?

File: Helper/Cart.php [hunk 3]
```diff
@@ -1304,7 +1304,7 @@ protected function getCalculationAddress($quote)
      * @return array
      * @throws \Exception
      */
-    public function getCartData($paymentOnly, $placeOrderPayload, $immutableQuote = null)
+    public function getCartData($paymentOnly, $placeOrderPayload, $immutableQuote = null, $requireBillingAddress = true)
```
Comment (vitaliyreznikov) [inline Helper/Cart.php:1307]: Can you update phpdoc?

Comment (danac-gs) [inline Helper/Cart.php:1307]: @dumega Taking into account the complexity of the problem and it's root case on Bolt end - not calling shipping and tax for virtual carts and inconsistency in processing guest vs. logged in user - discussed in our private slack channel, and having in mind that this PR only solves an edge case, leaving the main problem open, I'd like to suggest one simplification here that you may or may not consider.

From Bolt perspective there's only one address, `shipping == billing`. Although billing address is not present in every request type data, it is the same. So, I'll remove logic around `$requireBillingAddress` (everywhere) and use shipping address as billing one everywhere billing is missing. I've checked how `$requireBillingAddress` is used in the code and it does not make much sense.

File: Model/Api/DiscountCodeValidation.php [hunk 4]
```diff
@@ -800,7 +790,7 @@ private function getCartTotals($quote)
     {
         $request = $this->getRequestContent();
         $is_has_shipment = isset($request->cart->shipments[0]->reference);
```
Comment (danac-gs) [inline Model/Api/DiscountCodeValidation.php:792]: Note: shipments is populated for the user logged in Bolt checkout (shipping and tax is called every time in the process of generating checkout data) and also for the guest users after they apply the discount in the modal (shipping and tax gets called by Bolt as a side effect of discount code validation). No matter if the cart is virtual or not. However shipping and tax endpoint is not called for guest user virtual carts in general (no coupon applied). This is inconsistency in Bolt functionality that causes issues and should be fixed on Bolt end.

**Comment by dumega** [comment]
@danac-gs I think it is a plugin issue, not related to Bolt side, also it is not a edge case, but a common issue inside the plugin.

Let me show you an example, https://m2.nvip.eu/index.php/test-virtual.html this is a virtual product in my dev site, and for debugging, it can trigger the shipping&tax endpoint (the shipping and tax endpoint can be called for both guest user and logged-in user), so in the "Payment" step when we apply coupon on Bolt modal, the api request data of discounts.code.apply contains full billing address and shipments info. In this way, I think we can simulate the proper implement of applying coupon via Bolt modal.

And as we can see, the api hook still fails due to error that the billing address is empty (https://github.com/BoltApp/bolt-magento2/blob/master/Helper/Cart.php#L1478). So my point is, even Bolt fixes shipping and tax endpoint for virtual cart of guest user, this issue would still exist, that's why I says this is a common plugin issue. It is totally a different issue from the one  we discuss on slack (https://boltpay.slack.com/archives/CEEAATB2Q/p1592880744154800)

Although there's actually only one address in Bolt side, but the api hook send billing address and shipping info separately, and in the plugin side, we should always use billding address for checkout, that seems to be more logical in M2, so I think it is reasonable to check $requireBillingAddress for virtual cart.

**Comment by danac-gs** [comment]
@dumega The "virtual" product, https://m2.nvip.eu/index.php/test-virtual.html,  does not seem to be configured correctly. The type of the item is "unknown". Should be "digital".
Moreover, I think using shipping address as a fallback for billing one (only if billing is missing, of course) and removing `$requireBillingAddress` is more clear and more logical solution. First, we should prefer removing code instead of adding it, thus keeping the overall complexity lower if possible. Then, it complies with Bolt logic of having one address only. At least one should always be present in the request from Bolt. If there's no address (no billing nor shipping) then and error should happen.

**Comment by dumega** [comment]
@danac-gs The misconfigured product is intended, cause only using it, we can properly reproduce the issue with Bolt PPC, cause due to another issue on Bolt side (https://boltpay.atlassian.net/browse/EN-238), the shipping&tax endpoint wont be triggered for virtual cart.  And when applying coupon in Payment step of Bolt modal, the cart can be treated as virtual cart in M2 normally.

Well, my proposal is to wait until the virtual cart tax issue is fixed on Bolt side, then we can look into this issue again.

**Comment by vitaliyreznikov** [comment]
@dumega Why we want to postpone this until the virtual item tax issue is resolved?
Let's say use applies a discount on the first step of guest checkout. We still have the issue in this case, correct?

**Comment by dumega** [comment]
@vitaliyreznikov Yes, in my opinion, this is a plugin issue, whatever step you apply the discount, the error would happen. But @danac-gs  insist that the root cause is inconsistency in Bolt functionality and should be fixed on Bolt end. Then we have to wait until the virtual item tax issue is resolved and see if the issue still exist or not.

**Comment by danac-gs** [comment]
> @vitaliyreznikov Yes, in my opinion, this is a plugin issue, whatever step you apply the discount, the error would happen. But @danac-gs insist that the root cause is inconsistency in Bolt functionality and should be fixed on Bolt end. Then we have to wait until the virtual item tax issue is resolved and see if the issue still exist or not.

@vitaliyreznikov @dumega Virtual carts do not work. Yet. It's a known issue and it's on Bolt end. They might work in some really edge cases, for instance when the discount is applied at some (last?) step due to the circumstance that shipping and tax is also called in the process (at the end of) discount validation step. There's no point in fixing this now and making plugin work with virtual carts in this edge case only, before the main issue is fixed. After it's fixed, this probably won'e be an issue anymore.

**Comment by vitaliyreznikov** [comment]
@danac-gs What if we have a cart with virtual cart and apply a discount on the first step:
1) do we have an issue now?
2) can it be fixed on bolt side?

**Comment by danac-gs** [comment]
> @danac-gs What if we have a cart with virtual cart and apply a discount on the first step:
> 
> 1. do we have an issue now?
> 2. can it be fixed on bolt side?

@vitaliyreznikov Just to be sure we are all on the same page. All the above discussion refers only to virtual carts, a cart with ALL items virtual. @dumega to confirm
I don't think the aforementioned case, or any other case, including the one from the task, is relevant until the `virtual cart collect tax` issue is fixed on Bolt side.
Back to your question. With the update / fix on Bolt side, all the carts, virtual or physical, would (should) behave the same, with no issue of this type.
My point, again, is that we can not consider and present (to merchants) this as a solution because it is a fix for only a small portion of virtual carts, the ones that have discounts applied (at some or any checkout step, does't matter). On the other hand, this would be no issue with Bolt side fix applied. It is misleading for everyone - engineers, engineer management, customer support, merchants.

**Comment by vitaliyreznikov** [comment]
Let's close this for now, taking in account @danac-gs opinion

**Comment by dumega** [comment]
@vitaliyreznikov When does the team fix the virtual cart collect tax issue on Bolt side?

**Comment by vitaliyreznikov** [comment]
**Reopening:**
By the contract, we shouldn’t update cart (quote) on shipping&tax call. We need just provide information.
For example, for now we are working on static shipping project, when we miss shipping call because we know shipping price.
So long-term solution: don’t save billing address in quote
short-term: make sure that we are ok even if billing address is empty

**Comment by adentong** [comment]
This has been open for half year now... I'm just going to close this.
